### PR TITLE
Porting spark

### PR DESCRIPTION
**Pull request by stefano81**

### CODE REVIEW

File: library/build.gradle [hunk 1]
```diff
@@ -13,7 +13,7 @@ ext {
     junit_jupiter_version = '5.9.1'
     junit_platform_launcher_version = '1.9.1'
 
-    mockito_version = '4.11.0'
+    mockito_version = '4.10.0'
```
Comment (marcosimioni) [inline library/build.gradle:16]: why this downgrade? was this on purpose?

File: docs/spark/README.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+## Data Privacy Toolkit Spark Wrappers
+
+What follows describes the spark wrappers available to leverage the functionality of **DPT** in a Spark environment.
+
+The Spark JAR was built for Java 11 and Spark 3.2.3.
```
Comment (marcosimioni) [inline docs/spark/README.md:5]: is this still accurate?

File: docs/spark/risk-estimation.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
+## Invoking the Risk Estimation Spark executor
+
+The Spark JAR was built for Java 1.8 and Spark 2.2. The toolkit is packaged together with its dependencies so there is no requirement for additional configuration of classpaths or internet connection to artifact repositories.
```
Comment (marcosimioni) [inline docs/spark/risk-estimation.md:3]: this is not consistent with what we said in https://github.com/IBM/data-privacy-toolkit/pull/101/files#diff-a80b1b3e7affdb884c33245ed1fca2f940c94a451d90b53230d4dbf013360d2aR5 (see my previous comment)

File: library/src/main/java/com/ibm/research/drl/dpt/anonymization/PrivacyConstraint.java [hunk 4]
```diff
@@ -6,6 +6,7 @@
 package com.ibm.research.drl.dpt.anonymization;
 
 
+import com.fasterxml.jackson.annotation.JsonIgnore;
```
Comment (marcosimioni) [inline library/src/main/java/com/ibm/research/drl/dpt/anonymization/PrivacyConstraint.java:9]: shouldn't we upgrade the copyright to 2022 here?

File: library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/DistinctLDiversity.java [hunk 5]
```diff
@@ -103,6 +103,11 @@ public void initialize(IPVDataset dataset, List<ColumnInformation> columnInforma
         sanityChecks(dataset);
     }
 
+    @Override
```
Comment (marcosimioni) [inline library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/DistinctLDiversity.java:106]: same here, should upgrade copyright to 2022?

File: library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/EntropyLDiversity.java [hunk 6]
```diff
@@ -107,4 +107,9 @@ public void initialize(IPVDataset dataset, List<ColumnInformation> columnInforma
         sanityChecks(dataset);
     }
 
+    @Override
```
Comment (marcosimioni) [inline library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/EntropyLDiversity.java:110]: same here, should upgrade copyright to 2022?

File: library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/KAnonymity.java [hunk 7]
```diff
@@ -62,4 +62,9 @@ public void sanityChecks(IPVDataset originalDataset) {
     public void initialize(IPVDataset dataset, List<ColumnInformation> columnInformationList) {
         sanityChecks(dataset);
     }
+
```
Comment (marcosimioni) [inline library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/KAnonymity.java:65]: same here, should upgrade copyright to 2022?

File: library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/LDiversityMetric.java [hunk 8]
```diff
@@ -28,9 +28,9 @@ public LDiversityMetric(List<String> sensitiveValues) {
         super(1);
         this.histograms = new ArrayList<>();
 
-        for (int i = 0; i < sensitiveValues.size(); i++) {
+        for (String sensitiveValue : sensitiveValues) {
```
Comment (marcosimioni) [inline library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/LDiversityMetric.java:31]: same here, should upgrade copyright to 2022?

File: library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/RecursiveCLDiversity.java [hunk 9]
```diff
@@ -115,4 +115,9 @@ public void initialize(IPVDataset dataset, List<ColumnInformation> columnInforma
         sanityChecks(dataset);
     }
 
+    @Override
```
Comment (marcosimioni) [inline library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/RecursiveCLDiversity.java:118]: same here, should upgrade copyright to 2022?

not gonna create this comment for all of them, maybe we should doublecheck all the files we've touched for their copyright

**Review by marcosimioni — CHANGES_REQUESTED** [review]
LGTM apart from a couple of minor things, can you look at them before approving plz?

File: docs/spark/README.md (hunk 2 above)
Comment (stefano81) [reply docs/spark/README.md:5]: Yes, Spark version has been improved to the stable one for 3.2 but Java is maintained to 11 for consistency with previous engagements.

File: library/build.gradle (hunk 1 above)
Comment (stefano81) [reply library/build.gradle:16]: Thank you for pointing it out, uniforming to the latest version through the entire project

File: docs/spark/risk-estimation.md (hunk 3 above)
Comment (stefano81) [reply docs/spark/risk-estimation.md:3]: Requires update, missed it

File: library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/DistinctLDiversity.java (hunk 5 above)
Comment (stefano81) [reply library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/DistinctLDiversity.java:106]: No real changes to this file

File: library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/EntropyLDiversity.java (hunk 6 above)
Comment (stefano81) [reply library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/EntropyLDiversity.java:110]: No real changes to this file

File: library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/KAnonymity.java (hunk 7 above)
Comment (stefano81) [reply library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/KAnonymity.java:65]: No real changes to this file

File: library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/RecursiveCLDiversity.java (hunk 9 above)
Comment (stefano81) [reply library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/RecursiveCLDiversity.java:118]: No real changes to this file

File: library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/LDiversityMetric.java (hunk 8 above)
Comment (stefano81) [reply library/src/main/java/com/ibm/research/drl/dpt/anonymization/constraints/LDiversityMetric.java:31]: No real changes to this file

**Review by marcosimioni — APPROVED** [review]
LGTM

**Pull request merged**